server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1384,7 +1384,7 @@ public Metadata build() {
 
                     IndexMetadata writeIndex = backingIndices.get(backingIndices.size() - 1);
```

Comment:
I agree that the backing index list should have a defined order and that we can simplify the validation that the generation matches the last name in the list. 

If you're also suggesting that we keep the separate `writeIndex` parameter on the `IndexAbstraction.DataStream` constructor, I don't see why that would continue to be necessary since the value for the `writeIndex` parameter would always be `backingIndices.get(backingIndices.size() - 1)`?